### ISSUE TITLE
Training Hugging Face GPT2 on a TPU fails with ParallelLoader

### ISSUE THREAD

**Issue by roeischuster** [pytorch/xla#1931]
## ❓ Questions and Help

Hi all,

I ported the [language model training code of Hugging Face](https://github.com/huggingface/transformers/blob/master/examples/run_language_modeling.py) to work on XLA, simply following the instructions [here](https://pytorch.org/xla/release/1.5/index.html), and trying to train GPT2 using a TPU, on Pytorch v1.5.

Training on a single XLA device works, and is faster than your average GPU, but not as fast as I need.

Once I try loading the data using `torch_xla.distributed.parallel_loader.ParallelLoader` (in order to train on multi-core), it crashes right after the first batch. First, it produces the following warning on the first call to `xm.optimizer_step`:

> /pytorch/torch/csrc/utils/python_arg_parser.cpp:749: UserWarning: This overload of add_ is deprecated:
>         add_(Number alpha, Tensor other)
> Consider using one of the following signatures instead:
>         add_(Tensor other, Number alpha)
>   TORCH_WARN_ONCE(msg);

Then, it keeps running until the next iteration. When trying to get the next batch from the ParallelLoader, the following warning occurs, followed by a crash.

>  W   13773 tensorflow/core/distribut
> ed_runtime/rpc/grpc_remote_master.cc:160] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1587239630.911328540","description":"Error received from peer ipv4:10.90.155.74:8470","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC

Then, a dump ([attached](https://github.com/pytorch/xla/files/4497544/bug_report.log)) is produced, ending in a stack trace:

> terminate called after throwing an instance of 'std::runtime_error'
>   what():  tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1079 : Check failed: session->session()->Run( feed_inputs, {}, {cached_node.operations[0]}, &outputs) == ::tensorflow::Status::OK() (Aborted: Session 31c6a5eb1dcf0556 is not found. vs. OK)
> *** Begin stack trace ***
>         tensorflow::CurrentStackTrace[abi:cxx11]()
>         xla::XrtComputationClient::ReleaseHandles(std::vector<xla::XrtComputationClient::DeviceHandle, std::allocator<xla::XrtComputationClient::DeviceHandle> >*, std::function<xla::XrtSession::CachedNode const& (xla::XrtSession*, tensorflow::Scope const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&, xla::metrics::Metric*, xla::metrics::Counter*)
>         xla::XrtComputationClient::HandleReleaser()
>         xla::util::TriggeredTask::Runner()
>         clone
> *** End stack trace ***

The attached log is of a run wrapped in `xmp.spawn` with `nproc=1`. But the same happens when using the `ParallelLoader` even if I don't call `xmp.spawn` at all.

Anyone has an idea what's wrong? 

I also tried switching to pytorch-nightly, but that didn't work. Should I try 1.4?

**Comment by dlibenzi** [comment]
Any chance you can create a Colab repro based on this template?

https://github.com/pytorch/xla/blob/master/contrib/colab/issue-report.ipynb

**Comment by roeischuster** [comment]
Alright, eventually I found the fix.

The problem seems to be stemming from calling `zero_grad()` before calling `ParallelLoader`'s `next()` function. I.e., the following does **not** work:

```
epoch_iterator = pl.ParallelLoader(train_dataloader, [args.device]).per_device_loader(args.device)
model.zero_grad()
for step, batch in enumerate(epoch_iterator):
    TRAIN ITERATION
    model.zero_grad()
```

Whereas the following **does** work:
```
epoch_iterator = pl.ParallelLoader(train_dataloader, [args.device]).per_device_loader(args.device)
for step, batch in enumerate(epoch_iterator):
     model.zero_grad()
    TRAIN ITERATION
```

I am now curious to know why this happens!

**Comment by dlibenzi** [comment]
I tried with MNIST and does not happen, so must be a combo between your model and that usage.
Without a Colab repro for us it's hard to debug.
